config/reader.go

```diff
@@ -49,31 +58,149 @@ func getAgentInfo() (*apppb.AgentInfo, error) {
+func replacePlaceholdersInCloudConfig(config *Config) error {
+	packages := config.Packages
+
+	bytes, err := config.MarshalJSON()
+	if err != nil {
+		return errors.Wrap(err, "cannot marshal config to bytes")
+	}
+	updatedBytes, err := replacePackagePlaceholdersInFile(bytes, mapPlaceholderToRealPaths(packages))
+	if err != nil {
+		return errors.Wrap(err, "err replacing pathholders in config")
+	}
+
+	if err := config.UnmarshalJSON(updatedBytes); err != nil {
+		return errors.Wrap(err, "cannot unmarshal bytes into config")
+	}
+	return nil
+}
+
+// mapPlaceholderToRealPaths takes packages and generates expected placeholders for them.
+func mapPlaceholderToRealPaths(packages []PackageConfig) map[string]string {
+	// stores the placeholders <> what they need to be replaced to
+	packageMap := make(map[string]string, len(packages))
+	for _, p := range packages {
+		packageMap[getPackagePlaceholder(p)] = generateFilePath(p)
+	}
+
+	return packageMap
+}
+
+// GenerateConfigFromBytes converts a byte stream to a valid robot config
+// and replaces file placeholders as part of that conversion.
+func GenerateConfigFromBytes(filepath string, bytes []byte) (*Config, error) {
+	packages, err := getPackagesFromFile(bytes)
 	if err != nil {
 		return nil, err
 	}
-	defer utils.UncheckedErrorFunc(r.Close)
 
+	updatedBytes, err := replacePackagePlaceholdersInFile(bytes, mapPlaceholderToRealPaths(packages))
+	if err != nil {
+		return nil, errors.Wrap(err, "err replacing placeholders in config")
+	}
 	unprocessedConfig := &Config{
-		ConfigFilePath: "",
+		ConfigFilePath: filepath,
 	}
 
-	if err := json.NewDecoder(r).Decode(unprocessedConfig); err != nil {
-		// clear the cache if we cannot parse the file.
-		clearCache(id)
-		return nil, errors.Wrap(err, "cannot parse the cached config as json")
+	if err = unprocessedConfig.UnmarshalJSON(updatedBytes); err != nil {
+		return nil, errors.Wrap(err, "failed to decode config from json")
 	}
+
 	return unprocessedConfig, nil
 }
 
+// if the config is {package: orgID/name, type: module, version: 1} -> this will create a link of root/.viam/module/.data/orgID-name-1.
+func generateFilePath(config PackageConfig) string {
+	// first get the base root
+
+	// for backwards compatibility, packages right now don't have ml_models as a type.
+	// so if that is the case we need to join it right now to packages/../name
+	var dir string
+	if config.Type == "" {
+		// then this is not set and it must be an ml-model for backward compatibility -- but for now just join it without the type
+		// package manager will still create symlinks for these packages based on the packges path
+		dir = path.Clean(path.Join(viamDotDir, "packages", config.Name))
+	} else {
+		dir = path.Clean(path.Join(viamDotDir, "packages", GetPackageDirectoryFromType(config.Type), dataDir, HashName(config)))
+	}
+	return dir
+}
+
+func getPackagePlaceholder(config PackageConfig) string {
```

Comment:
formats what the placeholder replacement should look like based on the package config.